### PR TITLE
Action points and character system

### PR DESCRIPTION
**Pull request by noahknegt**
This PR adds a modular character system and the grid based movement with action points

### CODE REVIEW

File: Assets/Scripts/Core/GridPathfinding.cs [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+				// No path possible
+			}
+			ret.Reverse();
+
+			return ret;
+		}
+
+		private PathNode FindTarget(PathNode currentNode, PathNode targetNode) {
+			// Check the north node
+			if (currentNode.moveNorth) DetermineNodeValues(currentNode, currentNode.north, targetNode);
```
Comment (lokiy999) [inline Assets/Scripts/Core/GridPathfinding.cs:518]: Coding style: single line bodies of your if-statement go to a new line (this is the same for the follow three if-statements as well)

File: Assets/Scripts/Core/GridPathfinding.cs [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+				// Get the next node with the smallest F value
+				return GetSmallestFValueNode();
+			} else {
+				return null;
+			}
+		}
+
+		private void DetermineNodeValues(PathNode currentNode, PathNode testing, PathNode targetNode) {
+			// Dont work on null nodes
+			if (testing == null)
```
Comment (lokiy999) [inline Assets/Scripts/Core/GridPathfinding.cs:540]: Coding style: add "{" and "}" to the single-body if-statement

File: Assets/Scripts/Core/GridPathfinding.cs [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+
+			// Check to see if the node is the target
+			if (testing == targetNode) {
+				targetNode.parent = currentNode;
+				foundTarget_ = true;
+				return;
+			}
+
+			// Ignore Walls
+			if (currentNode.weight == WALL_WEIGHT || testing.weight == WALL_WEIGHT)
```
Comment (lokiy999) [inline Assets/Scripts/Core/GridPathfinding.cs:551]: Coding style: add "{" and "}" to the single-body if-statement

File: Assets/Scripts/Core/MapPos.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		public MapPos AddPosCopy(int x, int y) {
+			return new MapPos(this.x + x, this.y + y);
+		}
+
+		public static int Distance(MapPos p1, MapPos p2) {
+			return System.Math.Abs(p1.x - p2.x) + System.Math.Abs(p1.y - p2.y);
+		}
+		public static bool ListContains(List<MapPos> list, MapPos mapPos) {
+			// Check if map pos is in list
+			foreach (MapPos pos in list)
```
Comment (lokiy999) [inline Assets/Scripts/Core/MapPos.cs:59]: Coding style: add "{" and "}" to the foreach statement and the if-statement

File: Assets/Scripts/Core/MeshUtils.cs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	public static class MeshUtils {
+		
+		private static readonly Vector3 Vector3zero = Vector3.zero;
+		private static readonly Vector3 Vector3one = Vector3.one;
+		private static readonly Vector3 Vector3yDown = new Vector3(0,-1);
+
+		
+		private static Quaternion[] cachedQuaternionEulerArr;
+		private static void CacheQuaternionEuler() {
+			if (cachedQuaternionEulerArr != null) return;
```
Comment (lokiy999) [inline Assets/Scripts/Core/MeshUtils.cs:15]: Coding style: don't add a space in if-statements and for-loops in this file

File: Assets/Scripts/Core/MeshUtils.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace OperationBlackwell.Core {
+	public static class MeshUtils {
+		
+		private static readonly Vector3 Vector3zero = Vector3.zero;
+		private static readonly Vector3 Vector3one = Vector3.one;
+		private static readonly Vector3 Vector3yDown = new Vector3(0,-1);
```
Comment (lokiy999) [inline Assets/Scripts/Core/MeshUtils.cs:10]: Readability: Add a space between "," and "-1"

File: Assets/Scripts/Core/MeshUtils.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+			if (cachedQuaternionEulerArr != null) return;
+			cachedQuaternionEulerArr = new Quaternion[360];
+			for (int i=0; i<360; i++) {
+				cachedQuaternionEulerArr[i] = Quaternion.Euler(0,0,i);
+			}
+		}
+		private static Quaternion GetQuaternionEuler(float rotFloat) {
+			int rot = Mathf.RoundToInt(rotFloat);
+			rot = rot % 360;
+			if (rot < 0) rot += 360;
```
Comment (lokiy999) [inline Assets/Scripts/Core/MeshUtils.cs:24]: Coding style: single-lined if-statements go to a new line

File: Assets/Scripts/Core/MovementTilemap.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		public event EventHandler OnLoaded;
+		private Grid<TilemapObject> grid_;
+
+		public MovementTilemap(int width, int height, float cellSize, Vector3 originPosition) {
+			grid_ = new Grid<TilemapObject>(width, height, cellSize, originPosition, (Grid<TilemapObject> g, Vector3 v, int x, int y) => new TilemapObject(g, x, y));
+		}
+
+		public void SetTilemapSprite(Vector3 worldPosition, TilemapObject.TilemapSprite tilemapSprite) {
+			TilemapObject tilemapObject = grid_.GetGridObject(worldPosition);
+			if (tilemapObject != null) {
```
Comment (lokiy999) [inline Assets/Scripts/Core/MovementTilemap.cs:18]: Coding style: don't add a space to your if-statements in this file

File: Assets/Scripts/Core/MovementTilemapVisual.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			mesh_ = new Mesh();
+			GetComponent<MeshFilter>().mesh = mesh_;
+
+			Texture texture = GetComponent<MeshRenderer>().material.mainTexture;
+			float textureWidth = texture.width;
+			float textureHeight = texture.height;
+
+			uvCoordsDictionary_ = new Dictionary<MovementTilemap.TilemapObject.TilemapSprite, UVCoords>();
+
+			foreach (TilemapSpriteUV tilemapSpriteUV in tilemapSpriteUVArray_) {
```
Comment (lokiy999) [inline Assets/Scripts/Core/MovementTilemapVisual.cs:36]: Coding style: don't add a space after your foreach-loop

File: Assets/Scripts/Core/MovementTilemapVisual.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			if (updateMesh_) {
+				updateMesh_ = false;
+				UpdateHeatMapVisual();
+			}
+		}
+
+		private void UpdateHeatMapVisual() {
+			MeshUtils.CreateEmptyMeshArrays(grid_.GetWidth() * grid_.GetHeight(), out Vector3[] vertices, out Vector2[] uv, out int[] triangles);
+
+			for (int x = 0; x < grid_.GetWidth(); x++) {
```
Comment (lokiy999) [inline Assets/Scripts/Core/MovementTilemapVisual.cs:70]: Coding style: don't add a space after your for-loops

File: Assets/Scripts/Core/PathNode.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		public event EventHandler OnWalkableChanged;
+
+		public int xPos;
+		public int yPos;
+		public PathNode parent;
+		public PathNode north;
+		public PathNode south;
+		public PathNode west;
+		public PathNode east;
+		public bool moveNorth;
```
Comment (lokiy999) [inline Assets/Scripts/Core/PathNode.cs:18]: Suggestion: Not sure if we/you want to do this, but can't you add "moveNorth", "moveSouth", "moveWest" and "moveEast" to the same line and declare them all at once, the same for the 4 cardinal directions above. This saves lines, but I am not sure if this improves the readability of the code.

File: Assets/Scripts/Core/PathRoute.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+		public PathRoute(List<PathNode> pathNodeList, List<Vector3> pathVectorList, MapPos finalPos) {
+			this.pathNodeList = pathNodeList;
+			this.pathVectorList = pathVectorList;
+			this.finalPos = finalPos;
+		}
+
+		public PathRoute(List<PathNode> pathNodeList, Vector3 worldOrigin, float nodeSize, MapPos finalPos) {
+			this.pathNodeList = pathNodeList;
+			pathVectorList = new List<Vector3>();
+			foreach (PathNode pathNode in pathNodeList) {
```
Comment (lokiy999) [inline Assets/Scripts/Core/PathRoute.cs:22]: Coding style: don't add a space before your foreach-loop in this file

File: Assets/Scripts/Player/GridCombatSystem.cs [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+		}
+
+		private void Start() {
+			blueTeamList_ = new List<UnitGridCombat>();
+			redTeamList_ = new List<UnitGridCombat>();
+			blueTeamActiveUnitIndex_ = -1;
+			redTeamActiveUnitIndex_ = -1;
+
+			// Set all UnitGridCombat on their GridPosition
+			foreach (UnitGridCombat unitGridCombat_ in unitGridCombatArray_) {
```
Comment (lokiy999) [inline Assets/Scripts/Player/GridCombatSystem.cs:33]: Coding style: Don't add a space after your foreach-loop and if-statements in thie file

File: Assets/Scripts/Player/GridCombatSystem.cs [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+						}
+					} else {
+						// Position is not Walkable
+					}
+				}
+			}
+		}
+
+		private void Update() {
+			switch (state_) {
```
Comment (lokiy999) [inline Assets/Scripts/Player/GridCombatSystem.cs:127]: Coding style: Don't add a space after your switch-statement

File: Assets/Scripts/Player/MovePositionDirect.cs [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+			movePosition_ = transform.position;
+		}
+
+		public void SetMovePosition(Vector3 movePosition) {
+			this.movePosition_ = movePosition;
+		}
+
+		private void Update() {
+			Vector3 moveDir = (movePosition_ - transform.position).normalized;
+			if (Vector3.Distance(movePosition_, transform.position) < 1f) moveDir = Vector3.zero; // Stop moving when near
```
Comment (lokiy999) [inline Assets/Scripts/Player/MovePositionDirect.cs:19]: Coding style: Don't add a space after your if-statement, and single-lined bodies of if-statements go to a new line

File: Assets/Scripts/Player/MovePositionPathfinding.cs [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+		private int pathIndex_ = -1;
+
+		public void SetMovePosition(Vector3 movePosition, Action onReachedTargetPosition) {
+			this.onReachedTargetPosition_ = onReachedTargetPosition;
+			pathVectorList_ = GridPathfinding.Instance.GetPathRouteWithShortcuts(transform.position, movePosition).pathVectorList;
+			/*Debug.Log("##########");
+			foreach (Vector3 vec in pathVectorList_) {
+				Debug.Log(vec);
+			}*/
+			if (pathVectorList_.Count > 0) {
```
Comment (lokiy999) [inline Assets/Scripts/Player/MovePositionPathfinding.cs:20]: Coding style: Don't add spaces after your if-statements in this file

File: Assets/Scripts/Player/PlayerMovementKeys.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace OperationBlackwell.Player {
+	public class PlayerMovementKeys : MonoBehaviour {
+		private void Update() {
+			float moveX = 0f;
+			float moveY = 0f;
+
+			if (Input.GetKey(KeyCode.W)) moveY = +1f;
```
Comment (lokiy999) [inline Assets/Scripts/Player/PlayerMovementKeys.cs:11]: Coding style: Single line bodied if-statements go to a new line

File: Assets/Scripts/Player/PlayerMovementMouse.cs [hunk 18]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using OperationBlackwell.Core;
+
+namespace OperationBlackwell.Player {    
+	public class PlayerMovementMouse : MonoBehaviour {
+		private void Update() {
+			if (Input.GetMouseButtonDown(1)) {
```
Comment (lokiy999) [inline Assets/Scripts/Player/PlayerMovementMouse.cs:9]: Coding style: Don't add a space after your if-statements

File: Assets/Scripts/Player/UnitGridCombat.cs [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+		private void Awake() {
+			characterBase_ = GetComponent<PlayerBase>();
+			selectedGameObject_ = transform.Find("Selected").gameObject;
+			movePosition_ = GetComponent<MovePositionPathfinding>();
+			//SetSelectedVisible(false);
+			state_ = State.Normal;
+		}
+
+		private void Update() {
+			switch (state_) {
```
Comment (lokiy999) [inline Assets/Scripts/Player/UnitGridCombat.cs:34]: Coding style: Don't add a space after your switch-statements in this file

File: Assets/Scripts/Core/GameController.cs [hunk 20]
```diff
@@ -16,62 +21,33 @@ public class GameController : MonoBehaviour {
...
-			grid_ = new Grid<Tilemap.Node>((int)gridWorldSize_.x, (int)gridWorldSize_.y, nodeRadius_, new Vector3(0, 0, 0), 
+		private void Awake() {
+			grid_ = new Grid<Tilemap.Node>((int)gridWorldSize_.x, (int)gridWorldSize_.y, cellSize_, new Vector3(0, 0, 0), 
 				(Grid<Tilemap.Node> g, Vector3 worldPos, int x, int y) => new Tilemap.Node(worldPos, x, y, g, true, true, false));
 			tilemap_ = new Tilemap(grid_);
 			tilemap_.SetTilemapVisual(tilemapVisual_);
+			Instance = this;
+			Vector3 origin = new Vector3(0, 0);
+
+			gridPathfinding = new GridPathfinding(origin + new Vector3(1, 1) * cellSize_ * .5f, new Vector3(gridWorldSize_.x, gridWorldSize_.y) * cellSize_, cellSize_);
```
Comment (lokiy999) [inline Assets/Scripts/Core/GameController.cs:36]: Suggestion: Cut this line into two seperate lines for readability

**Review by lokiy999 — CHANGES_REQUESTED** [review]
A lot of coding style reworks needs to be done, also I'm wondering if all the commented out code (not the commented out debug-code) is required and could maybe be removed to improve the code and the readability, otherwise the code might become a mess over time.

File: Assets/Scripts/Core/GridPathfinding.cs [hunk 21]
```diff
@@ -0,0 +1,713 @@
...
+			int height = heightMax_;
+
+			startPos = GetClosestValidPos(startPos.x, startPos.y);
+
+			if(startPos.x < 0 || startPos.y < 0 || finalPos.x < 0 || finalPos.y < 0 ||
+				startPos.x >= width || finalPos.x >= width ||
+				startPos.y >= height || finalPos.y >= height) {
+				return null; //Out of bounds!
+			}
+			if(mapNodes_[finalPos.x][finalPos.y].weight == WALL_WEIGHT ||
```
Comment (lokiy999) [inline Assets/Scripts/Core/GridPathfinding.cs:272]: Coding style: Missing : "{" and "}" for the if-statement

File: Assets/Scripts/Core/GridPathfinding.cs [hunk 22]
```diff
@@ -0,0 +1,713 @@
...
+				// Get the next node with the smallest F value
+				return GetSmallestFValueNode();
+			} else {
+				return null;
+			}
+		}
+
+		private void DetermineNodeValues(PathNode currentNode, PathNode testing, PathNode targetNode) {
+			// Dont work on null nodes
+			if(testing == null)
```
Comment (lokiy999) [inline Assets/Scripts/Core/GridPathfinding.cs:570]: Coding style: Missing "{" and "}" with the if-statement

File: Assets/Scripts/Core/IGameObject.cs [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+using System;
+
+namespace OperationBlackwell.Core {
+	public interface IGameObject {
+		void MoveTo(Vector3 targetPosition, Action onReachedPosition);
+		Vector3 GetPosition();
+		Team GetTeam();
+		bool IsEnemy(IGameObject unitGridCombat);
+	}
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/IGameObject.cs:11]: Missing newline at the end of file.

File: Assets/Scripts/Core/MapPos.cs (hunk 4 above)
Comment (Dennisbonke) [reply Assets/Scripts/Core/MapPos.cs:59]: This is not resolved yet, please fix. Same goes for line 60.

File: Assets/Scripts/Core/MapPos.cs [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+				if (pos.Equals(mapPos))
+					return true;
+			return false;
+		}
+
+		public string Save() {
+			// Returns a string to be used in savefiles
+			string[] content = new string[]{
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/MapPos.cs:67]: Needs a space after `[]`, also needs spaces before and after the `+` sign.

File: Assets/Scripts/Core/MapPos.cs [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+			string[] content = save.Split(new string[] { "#MAPPOS#" }, System.StringSplitOptions.None);
+			x = int.Parse(content[0]);
+			y = int.Parse(content[1]);
+			offsetX = float.Parse(content[2]);
+			offsetY = float.Parse(content[3]);
+			straightToOffset = bool.Parse(content[4]);
+		}
+	}
+
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/MapPos.cs:92]: Missing newline at the end of file.

File: Assets/Scripts/Core/MeshUtils.cs [hunk 26]
```diff
@@ -0,0 +1,157 @@
+﻿using System.Collections;
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/MeshUtils.cs:1]: This entire file seems duplicated, most of the functions can be found in `Utils.cs`, and any missing functions should go there too.

File: Assets/Scripts/Core/PathNode.cs [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+		}
+		public void CalculateFValue() {
+			fValue = gValue + hValue;
+		}
+		public Vector3 GetWorldVector(Vector3 worldOrigin, float nodeSize) {
+			return worldOrigin + new Vector3(xPos * nodeSize, yPos * nodeSize);
+		}
+	}
+
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/PathNode.cs:81]: Missing newline at the end of file.

File: Assets/Scripts/Core/PathQueue.cs [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+		public PathQueue(int _startX, int _startY, int _endX, int _endY, GridPathfinding.OnPathCallback _callback) {
+			startX = _startX;
+			startY = _startY;
+			endX = _endX;
+			endY = _endY;
+			callback = _callback;
+		}
+	}
+
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/PathQueue.cs:23]: Missing newline at the end of file.

File: Assets/Scripts/Core/PathRoute.cs [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+			this.finalPos = finalPos;
+		}
+
+		public void AddVector(Vector3 vector) {
+			pathVectorList.Add(vector);
+		}
+
+	}
+
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/PathRoute.cs:34]: Missing newline at the end of file.

File: Assets/Scripts/Core/Team.cs [hunk 30]
```diff
@@ -0,0 +1,6 @@
+namespace OperationBlackwell.Core {
+	public enum Team {
+		Blue,
+		Red
+	}
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Core/Team.cs:6]: Missing newline at the end of file.

File: Assets/Scripts/Player/MoveTransformVelocity.cs [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+
+		public void Disable() {
+			this.enabled = false;
+		}
+
+		public void Enable() {
+			this.enabled = true;
+		}
+	}
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Player/MoveTransformVelocity.cs:32]: Missing newline at the end of file.

File: Assets/Scripts/Player/MoveVelocity.cs [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+		public void Disable() {
+			this.enabled = false;
+			rigidbody2D_.velocity = Vector3.zero;
+		}
+
+		public void Enable() {
+			this.enabled = true;
+		}
+	}
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Player/MoveVelocity.cs:34]: Missing newline at the end of file.

File: Assets/Scripts/Player/PlayerBase.cs [hunk 33]
```diff
@@ -0,0 +1,14 @@
...
+	/*
+	* Character Base Class
+	* 
+	*/
+	public class PlayerBase : MonoBehaviour, IGetPosition {
+		public Vector3 GetPosition() {
+			return transform.position;
+		}
+	}
+}
```
Comment (Dennisbonke) [inline Assets/Scripts/Player/PlayerBase.cs:14]: Missing newline at the end of file.

File: Assets/Scripts/Player/UnitGridCombat.cs [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+		private void Awake() {
+			characterBase_ = GetComponent<PlayerBase>();
+			selectedGameObject_ = transform.Find("Selected").gameObject;
+			movePosition_ = GetComponent<MovePositionPathfinding>();
+			//SetSelectedVisible(false);
+			state_ = State.Normal;
+		}
+
+		private void Update() {
+			switch(state_) {
```
Comment (Dennisbonke) [inline Assets/Scripts/Player/UnitGridCombat.cs:34]: Missing a default label.

File: Assets/Scripts/Player/GridCombatSystem.cs [hunk 35]
```diff
@@ -0,0 +1,181 @@
...
+						}
+					} else {
+						// Position is not Walkable
+					}
+				}
+			}
+		}
+
+		private void Update() {
+			switch(state_) {
```
Comment (lokiy999) [inline Assets/Scripts/Player/GridCombatSystem.cs:127]: Coding style: each switch statement needs a default case

**Review by Dennisbonke — APPROVED** [review]
LGTM now.

**Pull request merged**